### PR TITLE
Replace uses of templates by using an interface instead.

### PR DESCRIPTION
**Pull request by baldersheim**
None of these uses were necessary for speed.
It just complicated code generation.

@vekterli PR

### CODE REVIEW

File: vespamalloc/src/vespamalloc/malloc/datasegment.h [hunk 1]
```diff
@@ -8,19 +8,36 @@
 
 namespace vespamalloc::segment {
 
-template<typename MemBlockPtrT>
+class IMemblockInfo {
```
Comment (vekterli) [inline vespamalloc/src/vespamalloc/malloc/datasegment.h:11]: Nit: should this be `IMemBlockInfo` to match the previous `MemBlockPtrT` casing?

**Review by vekterli — APPROVED** [review]
:+1:

File: vespamalloc/src/vespamalloc/malloc/datasegment.h [hunk 2]
```diff
@@ -8,19 +8,36 @@
...
+    virtual std::unique_ptr<IMemblockInfo> createMemblockInfo(void * ptr) const = 0;
+};
+
 class DataSegment
 {
 public:
     DataSegment(const DataSegment & rhs) = delete;
     DataSegment & operator = (const DataSegment & rhs) = delete;
-    DataSegment() __attribute__((noinline));
+    DataSegment(const IHelper & helper) __attribute__((noinline));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespamalloc/src/vespamalloc/malloc/datasegment.h:33]: Fixed

File: vespamalloc/src/vespamalloc/malloc/datasegment.h (hunk 1 above)
Comment (baldersheim) [reply vespamalloc/src/vespamalloc/malloc/datasegment.h:11]: Fixed

**Pull request merged**